linalg/constraints.hpp

```diff
@@ -0,0 +1,404 @@
+
+   /** @brief Constructor, elimination is by blocks.
+
+       The nonzeros in B are assumed to be in disjoint rows and columns; the
+       rows are identified with the lagrange_rowstarts array, the secondary
+       dofs are assumed to be the first nonzeros in the rows. */
+   EliminationSolver(HypreParMatrix& A, SparseMatrix& B,
+                     Array<int>& lagrange_rowstarts);
+
+   ~EliminationSolver();
+
+   void PrimalMult(const Vector& x, Vector& y) const override;
+
+protected:
+   /// Internal utility routine; assembles eliminated matrix explicitly
+   void BuildExplicitOperator();
+
+   /// Build preconditioner for eliminated system
+   virtual Solver* BuildPreconditioner() const = 0;
+   /// Select krylov solver for eliminated system
+   virtual IterativeSolver* BuildKrylov() const = 0;
+
+   HypreParMatrix& hA;
+   Array<Eliminator*> eliminators;
+   EliminationProjection * projector;
+   HypreParMatrix * h_explicit_operator;
+   mutable Solver* prec;
+};
+
+
+/** EliminationSolver using CG and HypreBoomerAMG */
+class EliminationCGSolver : public EliminationSolver
+{
+public:
+   EliminationCGSolver(HypreParMatrix& A, SparseMatrix& B,
+                       Array<int>& lagrange_rowstarts,
+                       int dimension_=0, bool reorder_=false) :
+      EliminationSolver(A, B, lagrange_rowstarts),
+      dimension(dimension_), reorder(reorder_)
+   { BuildExplicitOperator(); }
+
+protected:
+   virtual Solver* BuildPreconditioner() const override
+   {
+      HypreBoomerAMG * h_prec = new HypreBoomerAMG(*h_explicit_operator);
+      h_prec->SetPrintLevel(0);
+      if (dimension > 0) { h_prec->SetSystemsOptions(dimension, reorder); }
+      return h_prec;
+   }
+
+   virtual IterativeSolver* BuildKrylov() const override
+   { return new CGSolver(GetComm()); }
+
+private:
+   int dimension;
+   bool reorder;
+};
+
+
+/** @brief Solve constrained system with penalty method; see ConstrainedSolver.
+
+    Only approximates the solution, better approximation with higher penalty,
+    but with higher penalty the preconditioner is less effective. */
+class PenaltyConstrainedSolver : public ConstrainedSolver
+{
+public:
+   PenaltyConstrainedSolver(HypreParMatrix& A, SparseMatrix& B,
+                            double penalty_);
+
+   PenaltyConstrainedSolver(HypreParMatrix& A, HypreParMatrix& B,
+                            double penalty_);
+
+   ~PenaltyConstrainedSolver();
+
+   void PrimalMult(const Vector& x, Vector& y) const override;
+
+protected:
+   void Initialize(HypreParMatrix& A, HypreParMatrix& B);
+
+   /// Build preconditioner for penalized system
+   virtual Solver* BuildPreconditioner() const = 0;
+   /// Select krylov solver for penalized system
+   virtual IterativeSolver* BuildKrylov() const = 0;
+
+   double penalty;
+   Operator& constraintB;
+   HypreParMatrix * penalized_mat;
+   mutable Solver * prec;
+};
+
+
+/** Uses CG and a HypreBoomerAMG preconditioner for the penalized system. */
```

Comment:
Could the iterative solvers and preconditioners be parameters instead of hard-coded?

Reply:
What I intend here is a real C++ interface where a user who wants different solvers and preconditioners implements another class (lines 300-331 are the entire implementation). For example `dimension` and `reorder` only make sense here for systems AMG - other solver/preconditioner combinations will also have their own particular options, and I don't want to handle all of those in a single class.